Source/DafnyRuntime/DafnyRuntime.cs

```diff
@@ -1154,7 +1154,7 @@ internal class ArraySequence<T> : Sequence<T> {
   internal class ConcatSequence<T> : Sequence<T> {
     // INVARIANT: Either left != null, right != null, and elmts's underlying array == null or
     // left == null, right == null, and elmts's underlying array != null
-    private ISequence<T> left, right;
+    private volatile ISequence<T> left, right;
     private ImmutableArray<T> elmts;
```

Comment:
Ah shoot. In that case I'm not totally sure we're safe here, but I'm okay deferring to @cpitclaudel to approve the PR

Reply:
This code is cursed, but I think @fabiomadge 's implementation is right, because of the release semantics of volatile writes (https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#1454-volatile-fields has an example that does basically the same as this code).